libp2p/transports/tcptransport.nim

```diff
@@ -12,262 +12,325 @@
-
-  return transport
-
-method start*(
-  self: TcpTransport,
-  addrs: seq[MultiAddress]) {.async.} =
-  ## listen on the transport
-  ##
-
-  if self.running:
-    warn "TCP transport already running"
-    return
-
-  await procCall Transport(self).start(addrs)
-  trace "Starting TCP transport"
-  trackCounter(TcpTransportTrackerName)
+    T: typedesc[TcpTransport],
+    flags: set[ServerFlags] = {},
+    upgrade: Upgrade,
+    connectionsTimeout = 10.minutes,
+): T {.public.} =
+  T(
+    flags: flags,
+    clientFlags:
+      if ServerFlags.TcpNoDelay in flags:
+        {SocketFlags.TcpNoDelay}
+      else:
+        default(set[SocketFlags])
+    ,
+    upgrader: upgrade,
+    networkReachability: NetworkReachability.Unknown,
+    connectionsTimeout: connectionsTimeout,
+  )
+
+method start*(self: TcpTransport, addrs: seq[MultiAddress]): Future[void] =
+  ## Start transport listening to the given addresses - for dial-only transports,
+  ## start with an empty list
+
+  # TODO remove `impl` indirection throughout when `raises` is added to base
+
+  proc impl(
+      self: TcpTransport, addrs: seq[MultiAddress]
+  ): Future[void] {.async: (raises: [transport.TransportError, CancelledError]).} =
+    if self.running:
+      warn "TCP transport already running"
+      return
 
-  for i, ma in addrs:
-    if not self.handles(ma):
-      trace "Invalid address detected, skipping!", address = ma
-      continue
+    trace "Starting TCP transport"
 
     self.flags.incl(ServerFlags.ReusePort)
-    let server = createStreamServer(
-      ma = ma,
-      flags = self.flags,
-      udata = self)
 
-    # always get the resolved address in case we're bound to 0.0.0.0:0
-    self.addrs[i] = MultiAddress.init(
-      server.sock.getLocalAddress()
-    ).tryGet()
+    var supported: seq[MultiAddress]
+    var initialized = false
+    try:
+      for i, ma in addrs:
+        if not self.handles(ma):
+          trace "Invalid address detected, skipping!", address = ma
+          continue
 
-    self.servers &= server
+        let
+          ta = initTAddress(ma).expect("valid address per handles check above")
+          server =
+            try:
+              createStreamServer(ta, flags = self.flags)
+            except common.TransportError as exc:
+              raise (ref TcpTransportError)(msg: exc.msg, parent: exc)
+
+        self.servers &= server
+
+        trace "Listening on", address = ma
+        supported.add(
+          MultiAddress.init(server.sock.getLocalAddress()).expect(
+            "Can init from local address"
+          )
+        )
+
+      initialized = true
+    finally:
+      if not initialized:
+        # Clean up partial success on exception
+        await noCancel allFutures(self.servers.mapIt(it.closeWait()))
+        reset(self.servers)
 
-    trace "Listening on", address = ma
+    try:
+      await procCall Transport(self).start(supported)
+    except CatchableError:
+      raiseAssert "Base method does not raise"
```

Comment:
Could you please elaborate on what we're trying to achieve here?

Reply:
in a future version of the code, the base methods should be refactored to add `raises` annotations so that this try/catch can be removed - until that is done, this except is here just to silence the compiler whic.